packages/expo-updates/ios/EXUpdates/Multipart/UpdatesMultipartStreamReader.swift

```diff
@@ -0,0 +1,170 @@
+//  Copyright © 2019 650 Industries. All rights reserved.
+
+import Foundation
+
+public typealias MultipartCallback = (_ headers: [String: Any]?, _ content: Data?, _ done: Bool) -> Void
+
+/**
+ * Fork of React Native's RCTMultipartStreamReader that doesn't necessarily
+ * expect a preamble (first boundary is not necessarily preceded by CRLF).
+ */
+public final class UpdatesMultipartStreamReader {
+  private enum Constants {
+    static let crlf = "\r\n"
+    static let bufferLength = 4 * 1024
+  }
+
+  private let stream: InputStream
+  private let boundary: String
+
+  public init(inputStream: InputStream, boundary: String) {
+    self.stream = inputStream
+    self.boundary = boundary
+  }
+
+  public func readAllParts(withCompletionCallback callback: @escaping MultipartCallback) -> Bool {
+    var chunkStart = 0
+    var bytesSeen = 0
+
+    // First delimiter doesn't necessarily need to be preceded by CRLF (boundary can be first thing in body)
+    let firstDelimiter = Data("--\(boundary)\(Constants.crlf)".utf8)
+    let restDelimiter = Data("\(Constants.crlf)--\(boundary)\(Constants.crlf)".utf8)
+    let closeDelimiter = Data("\(Constants.crlf)--\(boundary)--\(Constants.crlf)".utf8)
+
+    var delimiter = firstDelimiter
+    var content = Data()
+    var currentHeaders: [String: Any]?
+
+    var buffer = [UInt8](repeating: 0, count: Constants.bufferLength)
+
+    stream.open()
+    defer { stream.close() }
+
+    while true {
+      var isCloseDelimiter = false
+
+      // Search only a subset of chunk that we haven't seen before + few bytes
+      // to allow for the edge case when the delimiter is cut by read call
+      let searchStart = max(bytesSeen - closeDelimiter.count, chunkStart)
+      let remainingBufferRange = NSRange(location: searchStart, length: content.count - searchStart)
+
+      // Check for delimiters
+      var range = content.range(of: delimiter, options: [], in: remainingBufferRange)
+      if range.location == NSNotFound {
+        isCloseDelimiter = true
+        range = content.range(of: closeDelimiter, options: [], in: remainingBufferRange)
+      }
+
+      if range.location == NSNotFound {
+        if currentHeaders == nil {
+          parseHeadersIfFound(in: content, range: remainingBufferRange, chunkStart: chunkStart) { headers in
+            currentHeaders = headers
+          }
```

Comment:
optional: can make this function not use a callback (since we're already diverging from the fork anyways which is fine)

```suggestion
          if let newHeaders = parseHeadersIfFound(in: content, range: remainingBufferRange, chunkStart: chunkStart) {
            currentHeaders = newHeaders
          }
```